### PR TITLE
Rudy

### PR DESCRIPTION
**Pull request by albanD**

### CODE REVIEW

File: tree.cpp [hunk 1]
```diff
@@ -1,7 +1,77 @@
...
+            for(int source_label=0; source_label< nb_labels; ++source_label) {
+                edges[node_pos-1].weights[target_label][source_label] -= reparam_constant;
+            }
+        }
+    }
+
+    float dual_value = *min_element(nodes[0].unaries.begin(), nodes[0].unaries.end());
+    // Verify that all the min marginals agree and we didn't fuck up.
+    bool do_verification_check = false;
+    if(do_verification_check){
```
Comment (albanD) [inline tree.cpp]: I think this will always be false ;) (see one line above)

File: tree.cpp [hunk 2]
```diff
@@ -1,7 +1,77 @@
...
     this->nodes = nodes;
     this->edges = edges;
-}
\ No newline at end of file
+}
+
+float tree::forward_backward_min_marginals(){
+    int nb_nodes = nodes.size();
+
+    // Do the forward
```
Comment (albanD) [inline tree.cpp:15]: can we put this in a separate function? A very close variant can be used to compute the min marginals up to a certain variable no?
Or can we add the a node as parameter, in that case, we stop the forward at this node and stop the backward at this node hence giving us the min marginals at this node (only this node)

File: tree.cpp (hunk 1 above)
Comment (bunelr) [inline tree.cpp]: That's intended. If you want to diagnose, you can switch the flag. If you don't want to pay the overhead, you leave it to false.

File: tree.cpp (hunk 2 above)
Comment (bunelr) [inline tree.cpp:15]: so a function "forward until" and a "backward until", taking a node id as argument?

File: tree.cpp [hunk 3]
```diff
@@ -1,7 +1,106 @@
...
+}
+
+void tree::backward(int until_nodeId) {
+    int nb_nodes = nodes.size();
+    int node_pos = nb_nodes-1;
+    while(nodes[node_pos].id != until_nodeId) {
+        node from_node = nodes[node_pos];
+        edge linking_edge = edges[node_pos-1];
+
+        int nb_labels = from_node.unaries.size();
```
Comment (albanD) [inline tree.cpp:63]: @bunelr fyi you have a defined constant in problemStatic.h that contains this value: NBR_CLASSES

**Pull request merged**